File: src/freight_fate/models/career_training.py
"""Save-compatible company-driver training guidance."""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum

from .business import COMPANY_DRIVER, carrier_name
from .start_options import start_option


class TrainingStage(Enum):
    FIRST_DISPATCH = "first_dispatch"
    TRAINER_REMINDERS = "trainer_reminders"
    TRUST_OPENING = "trust_opening"
    TRUST_BUILDING = "trust_building"
    NORMAL_GUIDANCE = "normal_guidance"


@dataclass(frozen=True)
class TrainingGuidance:
    stage: TrainingStage
    title: str
    terminal_text: str
    dispatch_text: str
    recommendation_label: str

    @property
    def spoken_summary(self) -> str:
        return f"{self.title}. {self.terminal_text} {self.dispatch_text}"


def is_company_training_profile(profile) -> bool:
    return getattr(profile, "business_status", COMPANY_DRIVER) == COMPANY_DRIVER


def company_training_stage(profile) -> TrainingStage:
    deliveries = int(getattr(profile.career, "deliveries", 0))
    if deliveries <= 0:
        return TrainingStage.FIRST_DISPATCH
    if deliveries < 3:
        return TrainingStage.TRAINER_REMINDERS
    if deliveries == 3:
        return TrainingStage.TRUST_OPENING
    if deliveries < 10:
        return TrainingStage.TRUST_BUILDING
    return TrainingStage.NORMAL_GUIDANCE


def training_guidance(profile) -> TrainingGuidance:
    stage = company_training_stage(profile)
    carrier = carrier_name(profile)
    option = start_option(getattr(profile, "carrier_key", ""))
    flavor = _carrier_flavor(option.key, carrier)
    if stage is TrainingStage.FIRST_DISPATCH:
        return TrainingGuidance(
            stage,
            "First dispatch",
            f"{carrier} has you on real freight with trainer support close by.",
            f"{flavor} Start with a short standard load that leaves room on the appointment.",
            "trainer-recommended",
        )
    if stage is TrainingStage.TRAINER_REMINDERS:
        return TrainingGuidance(
            stage,
            "First-week service record",
            f"{carrier} is looking for steady service, not perfection, with trainer notes still close by.",
            f"{flavor} Favor short regional freight with clean timing.",
            "good first-week run",
        )
    if stage is TrainingStage.TRUST_OPENING:
        return TrainingGuidance(
            stage,
            "Dispatch trust opening",
            f"{carrier} has enough first-week history to widen the board.",
            "A reliable lane still helps your record more than chasing a difficult load.",
            "good lane to build your record",
        )
    if stage is TrainingStage.TRUST_BUILDING:
        return TrainingGuidance(
            stage,
            "Build dispatcher trust",
            f"{carrier} is watching on-time service, damage, and steady miles.",
            "Pick reliable lanes before chasing specialty freight.",
            "good service-record load",
        )
    return TrainingGuidance(
        stage,
        "Trusted company guidance",
        "Keep building seniority, clean service, endorsements, and better carrier lanes.",
        "Unlocked freight with good time margins is the strongest career move.",
        "trusted carrier lane",
    )


def training_recommendation_score(profile, job) -> float:
    stage = company_training_stage(profile)
    miles = float(getattr(job, "distance_mi", 0.0))
    deadline = float(getattr(job, "deadline_game_h", getattr(job, "deadline_h", 0.0)))
    margin = max(0.0, deadline - miles / 55.0)
    cargo = getattr(getattr(job, "cargo", None), "key", "")
    specialty_penalty = 60.0 if cargo in {"electronics", "machinery", "hazmat"} else 0.0

    if stage is TrainingStage.FIRST_DISPATCH:
        return miles + specialty_penalty - margin * 18.0
    if stage is TrainingStage.TRAINER_REMINDERS:
        return miles * 0.85 + specialty_penalty * 0.5 - margin * 14.0
    if stage in {TrainingStage.TRUST_OPENING, TrainingStage.TRUST_BUILDING}:
        return miles * 0.65 - margin * 20.0 + specialty_penalty * 0.25
    return miles


def _carrier_flavor(key: str, carrier: str) -> str:
    if key == "great_lakes_training":
        return f"{carrier} usually gives new hires extra appointment room."
    if key == "prairie_link":
        return f"{carrier} likes practical regional mileage."
    if key == "summit_value":
        return f"{carrier} rewards appointment discipline."
    return f"{carrier} keeps the first week balanced."

File: crates/ff-core/src/models/career_training.rs
//! Save-compatible company-driver training guidance (port of
//! `freight_fate/models/career_training.py`).

use crate::models::business_constants::COMPANY_DRIVER;
use crate::models::career::{carrier_name_of, CareerProfile, JobView};
use crate::models::start_options::start_option;

#[cfg(test)]
mod tests;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum TrainingStage {
    FirstDispatch,
    TrainerReminders,
    TrustOpening,
    TrustBuilding,
    NormalGuidance,
}

impl TrainingStage {
    /// The Python enum's `.value`.
    pub fn value(self) -> &'static str {
        match self {
            TrainingStage::FirstDispatch => "first_dispatch",
            TrainingStage::TrainerReminders => "trainer_reminders",
            TrainingStage::TrustOpening => "trust_opening",
            TrainingStage::TrustBuilding => "trust_building",
            TrainingStage::NormalGuidance => "normal_guidance",
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TrainingGuidance {
    pub stage: TrainingStage,
    pub title: String,
    pub terminal_text: String,
    pub dispatch_text: String,
    pub recommendation_label: String,
}

impl TrainingGuidance {
    fn new(
        stage: TrainingStage,
        title: &str,
        terminal_text: impl Into<String>,
        dispatch_text: impl Into<String>,
        recommendation_label: &str,
    ) -> Self {
        TrainingGuidance {
            stage,
            title: title.to_string(),
            terminal_text: terminal_text.into(),
            dispatch_text: dispatch_text.into(),
            recommendation_label: recommendation_label.to_string(),
        }
    }

    pub fn spoken_summary(&self) -> String {
        format!(
            "{}. {} {}",
            self.title, self.terminal_text, self.dispatch_text
        )
    }
}

pub fn is_company_training_profile<P: CareerProfile + ?Sized>(profile: &P) -> bool {
    profile.business_status() == COMPANY_DRIVER
}

pub fn company_training_stage<P: CareerProfile + ?Sized>(profile: &P) -> TrainingStage {
    let deliveries = profile.career().deliveries;
    if deliveries <= 0 {
        return TrainingStage::FirstDispatch;
    }
    if deliveries < 3 {
        return TrainingStage::TrainerReminders;
    }
    if deliveries == 3 {
        return TrainingStage::TrustOpening;
    }
    if deliveries < 10 {
        return TrainingStage::TrustBuilding;
    }
    TrainingStage::NormalGuidance
}

pub fn training_guidance<P: CareerProfile + ?Sized>(profile: &P) -> TrainingGuidance {
    let stage = company_training_stage(profile);
    let carrier = carrier_name_of(profile);
    let option = start_option(Some(profile.carrier_key()));
    let flavor = carrier_flavor(option.key, &carrier);
    match stage {
        TrainingStage::FirstDispatch => TrainingGuidance::new(
            stage,
            "First dispatch",
            format!("{carrier} has you on real freight with trainer support close by."),
            format!(
                "{flavor} Dispatch starts you on a short standard load with room on the appointment."
            ),
            "trainer-recommended",
        ),
        TrainingStage::TrainerReminders => TrainingGuidance::new(
            stage,
            "First-week service record",
            format!(
                "{carrier} is looking for steady service, not perfection, with trainer notes still close by."
            ),
            format!("{flavor} Expect short regional freight, and deliver it with clean timing."),
            "good first-week run",
        ),
        TrainingStage::TrustOpening => TrainingGuidance::new(
            stage,
            "Dispatch trust opening",
            format!("{carrier} has enough first-week history to widen the board."),
            "A reliable lane still helps your record more than chasing a difficult load.",
            "good lane to build your record",
        ),
        TrainingStage::TrustBuilding => TrainingGuidance::new(
            stage,
            "Build dispatcher trust",
            format!("{carrier} is watching on-time service, damage, and steady miles."),
            "Run reliable lanes before chasing specialty freight.",
            "good service-record load",
        ),
        TrainingStage::NormalGuidance => TrainingGuidance::new(
            stage,
            "Trusted company guidance",
            "Keep building seniority, clean service, endorsements, and better carrier lanes.",
            "Unlocked freight with good time margins is the strongest career move.",
            "trusted carrier lane",
        ),
    }
}

pub fn training_recommendation_score<P, J>(profile: &P, job: &J) -> f64
where
    P: CareerProfile + ?Sized,
    J: JobView + ?Sized,
{
    let stage = company_training_stage(profile);
    let miles = job.distance_mi();
    let deadline = job.deadline_game_h();
    let margin = (deadline - miles / 55.0).max(0.0);
    let cargo = job.cargo_key();
    let specialty_penalty = if matches!(cargo, "electronics" | "machinery" | "hazmat") {
        60.0
    } else {
        0.0
    };

    match stage {
        TrainingStage::FirstDispatch => miles + specialty_penalty - margin * 18.0,
        TrainingStage::TrainerReminders => miles * 0.85 + specialty_penalty * 0.5 - margin * 14.0,
        TrainingStage::TrustOpening | TrainingStage::TrustBuilding => {
            miles * 0.65 - margin * 20.0 + specialty_penalty * 0.25
        }
        TrainingStage::NormalGuidance => miles,
    }
}

fn carrier_flavor(key: &str, carrier: &str) -> String {
    if key == "great_lakes_training" {
        return format!("{carrier} usually gives new hires extra appointment room.");
    }
    if key == "prairie_link" {
        return format!("{carrier} likes practical regional mileage.");
    }
    if key == "summit_value" {
        return format!("{carrier} rewards appointment discipline.");
    }
    format!("{carrier} keeps the first week balanced.")
}
